### ISSUE TITLE
Animation interrupted causes a bug

### ISSUE THREAD

**Issue by Powkachu** [ashish-chopra/ngx-gauge#14]
Hello,

When we interrupt a gauge animation, for example when we navigate to another page before the animation ended, an error is displaying:
<img width="736" alt="screen shot 2018-02-28 at 14 16 38" src="https://user-images.githubusercontent.com/7096747/36789642-85f6c764-1c92-11e8-837f-d987bde8ba97.png">
I think the problem comes because the component (where the gauge is used) is destroyed and doesn't exist anymore.

**Comment by fernandocode** [comment]
You could create an example of the problem at https://stackblitz.com/fork/angular?

**Comment by victorj2307** [comment]
Same issue here, tried to reproduce on stackblitz but couldn't import NgxGaugeModule :disappointed:

My gauge is inside a modal and refresh the value using EventEmitter, nothing complicated.

HTML Code:
```
<ngx-gauge [type]="batchGaugeOpts.type" [cap]="batchGaugeOpts.cap" [thick]="batchGaugeOpts.thick" [value]="selectedBatch.progress"
                        [label]="batchGaugeOpts.label" [append]="batchGaugeOpts.append" [size]="batchGaugeOpts.size">
                    </ngx-gauge>
```

Component code:
```
    public batchGaugeOpts: any;
    public selectedBatch: any;

    constructor(
        private eventsService: EventsService,
        private modalService: NgbModal
    ) {
        this.batchGaugeOpts = {
            type: "arch",
            cap: "round",
            label: "Progress",
            append: "%",
            thick: 20,
            size: 300,
            value: 0
        };
        this.selectedBatch = null;
    }

    public openBatchDetail(batchResult: any): void {
        this.selectedBatch = {};
        this.selectedBatch = batchResult;
            if (batchResult.status == 'FINISHED'){
            this.selectedBatch.progress = 100
        }
        this.modalService.open(this.batchDetailModal, { size: 'lg' });
    }
```

**Comment by ashish-chopra** [comment]
Hi @victorj2307 

Try to add the exact version of ngx-gauge@1.0.0-beta.2 in stackblitz and after that you will be able to import NgxGaugeModule. Since it is in beta right now you have to specify exact version during npm install in Stackblitz.

@Powkachu, @victorj2307  I will look into the issue.

**Comment by fernandocode** [comment]
@victorj2307 , If you want to set the example, I have prepared a [basic Stackblitz with ngx-gauge configured](https://stackblitz.com/edit/angular-kaonzk).

**Comment by victorj2307** [comment]
Hi, finally I set the example on Stackblitz (https://stackblitz.com/edit/angular-yncdvp)

There is a "Detail" button you should click to open a bootstrap modal, the modal contains the Gauge. To reproduce the bug you should open de modal multiple times and like the third time the error appears in the Browser Console. I sometimes also found the error when I closed the modal by clicking outside the modal and not using the Close button.

Thanks for your help.

**Comment by fernandocode** [comment]
I found the probable problem.

The gauge after being destroyed can still be called by the animation, and the context canvas has no value since it is invalid in the `_destroy()` method.

So what I did, was to test the `_context` (`if (this._context)`) before trying to access its attributes and methods. So when attempting to use `_context `after it has already been invalidated, nothing will be accomplished and the error will be avoided.

[Follow this update in the example with the tests in `_context`, and see if this covers the problem.](https://stackblitz.com/edit/angular-jnrfxy?file=libs/ngx-gauge/src/gauge/gauge.ts)

If this raises the problem, we can prepare a PR with this change to be corrected in the next version of the application.

**Comment by victorj2307** [comment]
I just finished testing your changes (test _context) and it solves the problem.
I'll wait for the changes to be released in a new version.

Thank you for your support.

**Comment by fernandocode** [comment]
@ashish-chopra , where should I send the PR with this change, from the `master `branch or `feature/gauge-directives`?

And analyzing the existing PR, something has already been submitted with an attempt to cover this #15 , except that it is not checking everywhere where `_context `is being used.

**Comment by ashish-chopra** [comment]
Hi @fernandocode, if you wanted to send a PR, create a branch from master and send a PR there. Always follow this convention.

Secondly, Can you tell me where all you did the changes in the ngx-gauge code regarding checking the context. I am assuming you only modified _clear() method. Am I right?

Also, I am not sure, why i wrote `this._clear()` inside `ngOnDestroy()`. Quite strange to me!!  But what I feel here is we first need to destroy the running animation then we will clear the context. I may remove the `_clear()` call altogether.

What is your suggestion?

**Comment by ashish-chopra** [comment]
Hi @fernandocode and @victorj2307 

Thanks for creating the stackblitz with exact problem statement. I now able to test my changes blazingly fast. Checkout my version of changes as I mentioned in the previous comment here: https://stackblitz.com/edit/angular-kuveqx

Let me know your feedback, I will incorporate these changes then!

**Comment by fernandocode** [comment]
@ashish-chopra, I think this is the correct problem because it removes the cause of the problem, the previous solution was a "remedy for the disease" that is the "vaccine for disease", :laughing:

**Comment by victorj2307** [comment]
Hi @ashish-chopra 
I tested your solution and it's working fine for me.

Thanks for your solution and support.

**Comment by ashish-chopra** [comment]
Hello Everyone, 

The issue has been fixed and merged into master and published as ngx-gauge@1.0.0-beta.3
You can update your packages. the major version release is coming shortly in few days.

Thank you all for your collaborative effort in reproducing and solving the issue.

Commit ID: 96fdb1a3839148942cfca63732c1ba1c9b631666

And this is the updated stackblitz using the ngx-gauge package from npm. https://stackblitz.com/edit/angular-kuveqx

Please verify the change in your project! Thanks.

**Comment by victorj2307** [comment]
@ashish-chopra Tested and working fine :wink:
